### ISSUE TITLE
Issues running DeepCoord

### ISSUE THREAD

**Issue by burnCalories** [RealVNF/DeepCoord#12]
hi @stefanbschneider  
Due to lack of equipment, I use personal Windows laptop to build the project environment. I finished the initial pip install -r requirements.txt .Because the command is different, I did not use README setup and has passed tests.py
I try to run rlsp\agents\main.py command

if __name__ == '__main__':
    agent_config = 'D:/code/DeepCoord/res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'
    network = 'D:/code/DeepCoord/res/networks/abilene/abilene-in4-rand-cap0-2.graphml'
    service = 'D:/code/DeepCoord/res/service_functions/abc.yaml'
    sim_config = 'D:/code/DeepCoord/res/config/simulator/rand-mmp-arrival12-8_det-size001_dur100.yaml'
    cli([agent_config, network, service, sim_config, '70', '--append-test'])

But I got the following error

Using TensorFlow backend.
Traceback (most recent call last):
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 378, in <module>
    cli([agent_config, network, service, sim_config, '10', '--append-test'])
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 53, in cli
    append_test, best, sim_seed, gen_scenario)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 73, in setup
    setup_files(agent_helper, best)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 214, in setup_files
    logger = setup_logging(agent_helper.verbose, agent_helper.logfile)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 292, in setup_logging
    logging.config.fileConfig('logging.conf', disable_existing_loggers=False)
  File "C:\Anaconda\envs\distributed-drl-coordination-main\lib\logging\config.py", line 76, in fileConfig
    formatters = _create_formatters(cp)
  File "C:\Anaconda\envs\distributed-drl-coordination-main\lib\logging\config.py", line 109, in _create_formatters
    flist = cp["formatters"]["keys"]
  File "C:\Anaconda\envs\distributed-drl-coordination-main\lib\configparser.py", line 959, in __getitem__
    raise KeyError(key)
KeyError: 'formatters'

It seems that this is a parameter problem, but I don't know how to modify it .
I hope to successfully run deepcoord and carry out my further research. :)

**Comment by stefanbschneider** [comment]
Hi @burnCalories , happy to hear that you want to use DeepCoord. It should work with Windows out of the box, which is what I used locally when developing it. 

I think the problem is else where; seems like a known error message: https://stackoverflow.com/questions/23161745/python-logging-file-config-keyerror-formatters

From where did you run your command? From the `DeepCoord` project root folder, ie, the same one where `logging.conf` is located?

**Comment by stefanbschneider** [comment]
I tried to create a fix in this branch: https://github.com/RealVNF/DeepCoord/tree/fix-loading-logging-config 
Could you check out the branch (`git checkout fix-loading-logging-config`) and retry if the problem persists there?

**Comment by burnCalories** [comment]
hi @stefanbschneider , thanks for your help.
Today I made some other attempts. I found that the wrong boot file path was modified before causing an error 
for example  ”
 if __name__ == '__main__':
    agent_config = 'D:/code/DeepCoord/res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'
    network = 'D:/code/DeepCoord/res/networks/abilene/abilene-in4-rand-cap0-2.graphml'
    service = 'D:/code/DeepCoord/res/service_functions/abc.yaml'
    sim_config = 'D:/code/DeepCoord/res/config/simulator/rand-mmp-arrival12-8_det-size001_dur100.yaml'”
in D:\code\DeepCoord\src\rlsp\agents\main.py

When I change the path to
“ if __name__ == '__main__':
    agent_config = 'res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'
    network = 'res/networks/abilene/abilene-in4-rand-cap0-2.graphml'
    service = 'res/service_functions/abc.yaml'
    sim_config = 'res/config/simulator/rand-mmp-arrival12-8_det-size001_dur100.yaml'”
A new error has occurred 
“Using TensorFlow backend.
Usage: main.py [OPTIONS] AGENT_CONFIG NETWORK SERVICE SIM_CONFIG EPISODES
Try "main.py -h" for help.
Error: Invalid value for "AGENT_CONFIG": Path "Dres/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml" does not exist.”

And I also generated an error prompt when testing in the “test_gym_environment.py” file
Error
Traceback (most recent call last):
  File "D:\code\DeepCoord\src\tests\test_gym_environment.py", line 39, in test_min_max_delay
    config = get_config(AGENT_CONFIG)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 247, in get_config
    with open(config_file) as f:
FileNotFoundError: [Errno 2] No such file or directory: 'res/config/agent/sample_agent.yaml'

This seems to be caused by a file path error. At present, I haven't made any changes to the code of deepcoord. Do you have any suggestions? thanks ：）

**Comment by burnCalories** [comment]
I just used the fix code you released ，When I run with path 
“agent_config = 'res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml
 network = 'res/networks/abilene/abilene-in4-rand-cap0-2.graphml'
 service = 'res/service_functions/abc.yaml'
 sim_config = 'res/config/simulator/rand-mmp-arrival12-8_det-size001_dur100.yaml
 cli([agent_config, network, service, sim_config, '70', '--append-test'])''

Prompt error
“Using TensorFlow backend.
Usage: main.py [OPTIONS] AGENT_CONFIG NETWORK SERVICE SIM_CONFIG EPISODES
Try "main.py -h" for help.

Error: Invalid value for "AGENT_CONFIG": Path "res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml" does not exist.
”
When I use the full path，for example “agent_config = 'D:/code/DeepCoord/res/config/agent/ddpg/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'”

Prompt error
“Using TensorFlow backend.
Traceback (most recent call last):
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 389, in <module>
    cli([agent_config, network, service, sim_config, '70', '--append-test'])
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\code\DeepCoord\venv\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 55, in cli
    append_test, best, sim_seed, gen_scenario)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 75, in setup
    setup_files(agent_helper, best)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 225, in setup_files
    agent_helper.sim_config_path)
  File "D:\code\DeepCoord\src\rlsp\agents\main.py", line 285, in copy_input_files
    copyfile(agent_config_path, new_agent_config_path)
  File "D:\python\lib\shutil.py", line 121, in copyfile
    with open(dst, 'wb') as fdst:

FileNotFoundError: [Errno 2] No such file or directory: './results/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec/abilene-in4-rand-cap0-2/abc/rand-mmp-arrival12-8_det-size001_dur100/2023-02-14_21-39-26_seed9460/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'”

Thank you for taking the time to help me solve the problem ：）

**Comment by stefanbschneider** [comment]
So if you run on the new, fixed branch with the same input, it runs further and crashes with a different error than before? Ie, indicating that the previous error with loading the `logging.conf` is fixed in the branch? 
In that case, I'd merge the branch to master.

---

Looking further at the new errors, they all seem to be related to wrong paths or missing files. Did you check that all the files actually exist?
Was there any other output; did the agent train at all? 

> FileNotFoundError: [Errno 2] No such file or directory: './results/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec/abilene-in4-rand-cap0-2/abc/rand-mmp-arrival12-8_det-size001_dur100/2023-02-14_21-39-26_seed9460/agent_obs1_weighted-f0d0n1_64a_64c_099gam_00001tau_001alp_0001dec.yaml'”

Not sure what this is about. Maybe it's trying to load a trained agent? Can you check if the file exists? Can you try running the agent without `--append-test`?

**Comment by burnCalories** [comment]
@hi， @stefanbschneider   
About adding to the master, I don't know whether the full path of the agent_config/network/service file must be used for running because of the Windows system. I can't give you an accurate answer.
 I have an exciting news to analyze with you,now. When I run the test file, deepcoord runs successfully and creates the graph/results file. 
 " agent_config = 'D:/code/DeepCoord/res/config/agent/sample_agent.yaml'
    network = 'D:/code/DeepCoord/res/networks/sample_network.graphml'
    service = 'D:/code/DeepCoord/res/service_functions/abc.yaml'
    sim_config = 'D:/code/DeepCoord/res/config/simulator/sample_config.yaml'
    cli([agent_config, network, service, sim_config, '10', '--append-test'])"
Further tests were carried out to run another test run with the trained weights. successful !
     “cli([agent_config, network, service, sim_config, '70', '-t' , '2023-02-16_10-12-50_seed4902'])”

Since I just learned DRL, I don't know much about the generated files and their use. I want to ask you for some learning suggestions.
Finally, I will conduct further study and research. Thank you for your careful help

**Comment by stefanbschneider** [comment]
Does that mean your problem with running DeepCoord is solved for now? Happy to hear :) 

Could you summarize again what you did to resolve it?

---

Regarding learning suggestions for DRL in general:

* General intro: https://www.deepmind.com/learning-resources/introduction-to-reinforcement-learning-with-david-silver
* More practical: https://spinningup.openai.com/en/latest/
* More conceptual/theoretical: https://web.stanford.edu/class/psych209/Readings/SuttonBartoIPRLBook2ndEd.pdf

I'm sure there are a lot more resources that you can find through Google. Just start somewhere :)
